PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+        self.receiptFetcher = receiptFetcher
+        self.purchaserInfoManager = purchaserInfoManager
+        self.backend = backend
+        self.identityManager = identityManager
+        self.receiptParser = receiptParser
+        self.deviceCache = deviceCache
+    }
+
+    @objc public func restoreTransactions(completion maybeCompletion: ((PurchaserInfo?, Error?) -> Void)?) {
+        syncPurchases(receiptRefreshPolicy: .always, isRestore: true, maybeCompletion: maybeCompletion)
+    }
+
+    @objc public func syncPurchases(completion maybeCompletion: ((PurchaserInfo?, Error?) -> Void)?) {
+        syncPurchases(receiptRefreshPolicy: .never,
+                      isRestore: allowSharingAppStoreAccount,
+                      maybeCompletion: maybeCompletion)
+    }
+
+    @objc public func products(withIdentifiers identifiers: [String],
+                               completion: @escaping ([SKProduct]) -> Void) {
+        let productIdentifiersSet = Set(identifiers)
+        guard !productIdentifiersSet.isEmpty else {
+            operationDispatcher.dispatchOnMainThread { completion([]) }
+            return
+        }
+
+        productsManager.products(withIdentifiers: productIdentifiersSet) { products in
+            self.operationDispatcher.dispatchOnMainThread {
+                completion(Array(products))
+            }
+        }
+    }
+
+    @available(iOS 12.2, macOS 10.14.4, watchOS 6.2, macCatalyst 13.0, tvOS 12.2, *)
+    @objc public func paymentDiscount(forProductDiscount productDiscount: SKProductDiscount,
+                                      product: SKProduct,
+                                      completion: @escaping (SKPaymentDiscount?, Error?) -> Void) {
+        receiptFetcher.receiptData(refreshPolicy: .onlyIfEmpty) { maybeReceiptData in
+            guard let receiptData = maybeReceiptData,
+                  !receiptData.isEmpty else {
+                      completion(nil, ErrorUtils.missingReceiptFileError())
+                      return
+                  }
+            guard let discountIdentifier = productDiscount.identifier else {
+                      completion(nil, ErrorUtils.productDiscountMissingIdentifierError())
+                      return
+                  }
+            self.backend.post(offerIdForSigning: discountIdentifier,
+                              productIdentifier: product.productIdentifier,
+                              subscriptionGroup: product.subscriptionGroupIdentifier ,
+                              receiptData: receiptData,
+                              appUserID: self.appUserID) { maybeSignature, maybeKeyIdentifier, maybeNonce, maybeTimestamp, maybeError in
+                if let error = maybeError {
+                    // TODO: Consider making a custom error.
+                    completion(nil, error)
+                    return
+                }
+
+                guard let keyIdentifier = maybeKeyIdentifier,
```

Comment:
maybe this violates some separation of concerns, but i'm wondering whether checking for non-null response could be in backend?

Reply:
Yeah, I think that makes sense, ultimately we probably want to encapsulate these params into something like 
`SigningResponse` that holds all the outputs:

`maybeSignature, maybeKeyIdentifier, maybeNonce, maybeTimestamp`

I'll open an issue.

Reply:
https://github.com/RevenueCat/purchases-ios/issues/749